File: GeeksForGeeks/Math/Digits-in-Factorial.js
/**
 * Problem: Digits in Factorial
 * Platform: GeeksForGeeks
 * Difficulty: Easy
 *
 * Approach:
 * Sum log10(i) from 2 to n.
 * Number of digits = floor(sum) + 1.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

class Solution {
    digitsInFactorial(n) {
        if (n <= 1) {
            return 1;
        }

        let digits = 0;

        for (let i = 2; i <= n; i++) {
            digits += Math.log10(i);
        }

        return Math.floor(digits) + 1;
    }
}

File: GeeksForGeeks/Math/Digits-in-Factorial.py
"""
Problem: Digits in Factorial
Platform: GeeksForGeeks
Difficulty: Easy

Approach:
Instead of computing n!, use Kamenetsky's formula:

Digits = floor(log10(1) + log10(2) + ... + log10(n)) + 1

Time Complexity: O(n)
Space Complexity: O(1)
"""

import math

class Solution:
    def digitsInFactorial(self, n):
        if n <= 1:
            return 1

        digits = 0

        for i in range(2, n + 1):
            digits += math.log10(i)

        return math.floor(digits) + 1